### ISSUE TITLE
[RFC] Update years

### ISSUE THREAD

**Issue by chshersh** [vaclavsvejcar/headroom#29]
Often I have outdated years in my files, like `2017-2019`. It would be nice if the tool could update the latest year instead of replacing the default one, so I would have `2017-2020` in the header. Current behaviour: always puts `2020` (what is specified in the template).

**Comment by vaclavsvejcar** [comment]
To fit nicely the prepared changes (mainly for #25), I was thinking that this could be implemented using some kind of _dynamic variables_ (e.g. those not loaded from _YAML_ config or provided by command line option, but rather provided by _Headroom_ in runtime). Solution could be introducing new _dynamic variable_ `_current_year`, that would always contain the current year. Then the template header could look like:

```mustache
{-|
Module      : {{{ _haskell_module_name }}}
Description : {{{ _haskell_module_shortdesc }}}
Copyright   : (c) 2019-{{ _current_year }} {{ author }}
License     : {{ license }}
Maintainer  : {{ email }}
Stability   : experimental
Portability : POSIX

{{{ _haskell_module_longdesc }}}
-}
```

@chshersh Would that solve it for you?

**Comment by chshersh** [comment]
@vaclavsvejcar That definitely looks like an improvement 👍 But it won't work on every module. In some modules I have year range like `2017-2019` and some modules are newer, like `2019`, and some are already `2020`. So I can't just hardcode the previous year... To cover all possible cases, if the current year is 2020, then:

* `2020` should remain `2020`
* `2019` should become `2019-2020`
* `2019-2020` should remain `2019-2020`
* `2018-2019` should become `2018-2020`

Also, there are some cases when the module has copyrights for the previous authors. See an example here:

* https://github.com/kowainik/validation-selective/blob/master/src/Validation.hs

Ideally, it would be desirable to keep all previous authors the same without changes, to make the process fully automatic.

So, I guess, the desired behavior can be to have a dynamic variable `_updated_copyright`, so I can just use

```
Copyright   : {{ _updated_copyright }}
```

And get the desired behavior

**Comment by vaclavsvejcar** [comment]
Ah ok, got the idea. It's true that the current implementation isn't sufficient for such cases as different copyrights across modules. There's one thing though related to the solution based on `_updated_copyright` variable I'm not 100% sure about:

 __Default value:__
Using `{{ _updated_copyright }}` is fine as long you have extracted value from already existing _Haddock_ header, but what about the case there is no _Haddock_ header (e.g. adding license header to new module? Given the limitations of _Mustache_, the only feasible solution I know about is to probably use something as `{{ _updated_copyright }}{{^  _updated_copyright }} Copyright (c) {{ _current_year }} {{ author }} {{/ _updated_copyright }}` to explicitly state default value, which is probably fine, just little bit clumsy. Other option is to just keep the value of this field empty or put there some placeholder such as "!!! insert copyright here !!!" but that's not that comfortable.

@chshersh What's your opinion on this? Do you see any better solution how to deal with this case?

**Comment by chshersh** [comment]
@vaclavsvejcar Your solution seems okay to me. But I also I think it is one of those places where the default value will be okay. I don't think there's much variability in how `copyright` is specified by default.  So, if the default value of the `_updated_copyrigh` will be `(c) {{ _current_year }} {{ author }}`, this will be good enough for me.

**Comment by vaclavsvejcar** [comment]
@chshersh The only problem with such default value is that the `author` variable might not be present at all (_Headroom_ doesn't enforce you to define any required variables - `author` & friends are just selected for embedded template stubs and documentation to show the way how it could be done). On the other hand, allowing user to define the default value using the trick from my previous comment would require either to lift restriction that all values for variables in _Mustache_ template must be defined or to define some whitelist of variables not to check for presence, but that's ugly. I'll try to figure out some better solution.

**Comment by chshersh** [comment]
Another problem with having `{{ author }}` behind `{{ _updated_copyright }}` is a possible confusing error message. If the `author` variable is not defined, and `headroom` prints a message that `author` is  not defined, but user haven't used `author`, the message will be confusing. The message should tell that `author` is used implicitly in `_updated_copyright` to be more helpful.

**Comment by vaclavsvejcar** [comment]
@chshersh Here's quick summary about current progress on this task:

- Added the possibility to define variable values not only as plain strings, but also as templates itself, so you can compose variables easily (see below)
- Added the `_current_year` _dynamic variable_ that always contains the value of current year.
- Added the `_haddock_module_copyright_` _dynamic variable_ that contains unchanged value of `Copyright`field extracted from existing Haddock header.
- Added the `_haddock_module_copyright_updated` _dynamic variable_, which is the same as the above one, just with updated years. If no copyright is extracted from processed source code file, some default text will be placed as its value. You can override this by defining default value of this variable in _YAML_ configuration/cmdline argument, such as:
```yaml
_haskell_module_copyright_updated: "(c) {{ _current_year }} {{ author }}"
```

### What's not finished
I still need to fix broken indentation for multi-line fields such as `Copyright`, that is broken if the formatting differs between old and new header, e.g.

```mustache
Copyright         : {{{ _haddock_module_copyright_updated }}}
```
__VS__
```mustache
Copyright : {{{ _haddock_module_copyright_updated }}}
```

In that case, the result looks like this:

```haskell
{- |
Module                  : Summoner.Template.Script
Copyright               : (c) 2017-2020 Kowainik
           (c) 2008-2020 2nd Author
SPDX-License-Identifier : MPL-2.0
Maintainer              : Kowainik <xrom.xkov@gmail.com>
Stability               : Stable
Portability             : Portable

File templates for @cabal@ and @stack@ scripts.
-}
```

I'll need to figure out how to adjust this in newly generated headers, will let you know once this will be finished.

**Comment by chshersh** [comment]
@vaclavsvejcar Great progress! Keep up the good work :muscle: `headroom` is becoming more powerful and helpful :slightly_smiling_face: 

Regarding the updated years for the current author: I think you can update years for the author with the most recent years. This sounds legit to me since the _current author_ is the latest author in terms of time (in my understanding).

**Comment by vaclavsvejcar** [comment]
> Great progress! Keep up the good work 💪 `headroom` is becoming more powerful and helpful 🙂

@chshersh Thanks a lot 🙂 I'm trying to make _Headroom_ as best as I can and will happily add any new reasonable feature.

> Regarding the updated years for the current author: I think you can update years for the author with the most recent years. This sounds legit to me since the _current author_ is the latest author in terms of time (in my understanding).

That sounds good, but still there might be case when two (or more) authors would fit this condition, for example:

```haskell
{-|
Copyright               : (c) 2019 Kowainik
                          (c) 2019 2nd Author
-}
```

How should this be processed? Update both, update first one or last one?

**Comment by chshersh** [comment]
@vaclavsvejcar Hmm, maybe in that case you can compare _Author_ with the value of the `{{ author }}` variable?

**Comment by vaclavsvejcar** [comment]
@chshersh I'll probably need to go this way, although I'm not really satistied with such solution. The reason is that it expects some variable of some name to be defined. Also that amount of decisions if performs _under the hood_ is getting pretty big and maybe not such clear for end users (rules for deciding whether to update year + rules for deciding who's current author). Guess this feature is kinda must-have in order to use _Headroom_ for your projects, right?

I'll probably give it one more day thinking about any better solution and if I won't figure out anything better, I'll implement this one.

**Comment by vaclavsvejcar** [comment]
@chshersh So I was thinking about this and came up with probably much better and cleaner solution:

### Problems with `_haddock_module_copyright_updated` approach
I think this approach has some major flaws and would also require to implement some _not-so-nice_ piece of code:

1. __Deciding things for user__ - I don't really like piece of code that tries to decide important things for user, behind the scenes, and this is exactly the example. Behind the simple variable `_haddock_module_copyright_updated`, there are many decisions happening, e.g. what pattern actually update (`yyyy-yyyy` vs `yyyy - yyyy`), which copyright owner is the current one (and hence which one to update and which one not, assume presence of some `{{ author }}` variable). This could be explained in documentation, but still, I'd prefer solution that's more predictable.
1. __Hard to configure__ - As mentioned above, this approach would require to know who's the current author. It could fetch this for example from the `{{ author  }}` variable, but here's the problem that it assumes presence of some variable. And I really cannot figure out any way how to nicely expose configuration for this.
1. __Haddock-specific__ - This approach is also really specific to single _Haddock_ field (`Copyright`), even though this problem (of updating years) is really general and can be applied to  any part of any generated _license header_.

### Postprocess functions based approach
The solution I came up with is to completely scrap the idea about the `_haddock_module_copyright_updated` variable and introduce new higher-level approach, _post processing functions_. Such function is basically a function, that takes rendered license header, some configuration and performs some post processing on the header, before such header is placed into the source code file. IMO this approach as major advantages over the former one:
1. __General solution__ - it can be not only reused for other _Haddock_ fields, but also for completely different file types other than Haskell.
1. __Easy to configure__ - Each implemented post processing function can have its own configuration, so end user can easily adjust the behaviour to suit her/his needs.

### Example of updating years function
Consider this to be more overall example how it could look like, rather than as final solution, but let's say that the function updating years in generated license headers can be enabled using following configuration in `.headroom.yaml`:

```yaml
post-process:
  update-copyright:
    current-author-only: "{{ author }}" # yep you would be able to use variables here
```

What do you think? 🙂

**Comment by chshersh** [comment]
@vaclavsvejcar Your new approach also makes sense to me 👍 It is more flexible and covers more use-cases. I'm happy with it as well 🙂

**Comment by vaclavsvejcar** [comment]
@chshersh I implemented the copyright updating functionality as _postprocessing header function_, that can be enabled using following configuration:

```yaml
post-process:
  update-copyright:
    enabled: true
    config:
      selected-authors-only: ["Author to update"]
```
You can either specify the only authors to update using the `selected-authors-only`, or ommit this key at all and it will update all of them. You can also use existing variables as value, for example:

```yaml
variables:
  author: Kowainik
  email: xrom.xkov@gmail.com

...

post-process:
  update-copyright:
    enabled: true
    config:
      selected-authors-only: ["{{ author }}"]
```

Now _Headroom_ should support all you need to use it in your project, I'll past some more _Summoner_-specific info to the pull request - https://github.com/kowainik/summoner/pull/457.

**Comment by vaclavsvejcar** [comment]
Implemented and released as part of [v0.3.0.0](https://github.com/vaclavsvejcar/headroom/releases/tag/v0.3.0.0).

- https://doc.norcane.com/headroom/v0.3.0.0/whats-new/
